### PR TITLE
fix bug of generating enum class in python-fastapi generator

### LINKED ISSUE #10127: [BUG] python-fastapi generator problem with enumeration.
#### Bug Report Checklist

- [x] Have you provided a full/minimal spec to reproduce the issue?
- [x] Have you validated the input using an OpenAPI validator ([example](https://apidevtools.org/swagger-parser/online/))?
- [x] Have you [tested with the latest master](https://github.com/OpenAPITools/openapi-generator/wiki/FAQ#how-to-test-with-the-latest-master-of-openapi-generator) to confirm the issue still exists?
- [x] Have you searched for related issues/PRs?
- [x] What's the actual output vs expected output?
- [ ] [Optional] Sponsorship to speed up the bug fix or feature request ([example](https://github.com/OpenAPITools/openapi-generator/issues/6178))

##### Description

openapi-generator-cli -g python-fastapi generator cannot generate the valid enum class.

##### openapi-generator version

5.2.0

### PR DESCRIPTION
**Pull request by liulu1998**
Fix #10127
Previous `python-fastapi` generator generated enum classes with empty body.  In this PR, I try to generate Python's standard enum class, mapping values to `str`.

### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh
  ./bin/utils/export_docs_generators.sh
  ``` 
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  For Windows users, please run the script in [Git BASH](https://gitforwindows.org/).
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master` (6.3.0) (minor release - breaking changes with fallbacks), `7.0.x` (breaking changes without fallbacks)
- [x] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.  @spacether

### CODE REVIEW

File: modules/openapi-generator/src/main/resources/python-fastapi/model.mustache [hunk 1]
```diff
@@ -18,6 +25,21 @@ from pydantic import AnyUrl, BaseModel, EmailStr, Field, validator  # noqa: F401
 
 {{#models}}
 {{#model}}
+{{#isEnum}}
+class {{classname}}(str, Enum):
```
[the remark this answers is not in the document]
Comment (Linus-Boehm) [reply modules/openapi-generator/src/main/resources/python-fastapi/model.mustache:29]: Aren't enums in OpenApi not always string based?
https://swagger.io/docs/specification/data-models/enums/

The only way I found to generate integer like enums would be something like:

``` yaml
# Source: https://stackoverflow.com/questions/66465888/how-to-define-enum-mapping-in-openapi

Severity:
  type: integer
  oneOf:
    - title: HIGH
      const: 2
      description: An urgent problem
    - title: MEDIUM
      const: 1
    - title: LOW
      const: 0
      description: Can wait forever
```

But I think that would  need anyways a different handling, as the example is not even typed as enum.

Even this being maybe not perfect, it is already a big improvement to no generation at all.

**Comment by Linus-Boehm** [comment]
Great initiative! @liulu1998 are you working on the tests or do you need support?

**Comment by liulu1998** [comment]
@Linus-Boehm I'm working on those tests.

**Comment by maxiptah** [comment]
I've tried to generate a python-fastapi server now and get empty enum classes like this:
```
class DialogStartTypes(BaseModel):
    """NOTE"""
```
So this is I guess will be handled by this issue. Are you also aware that in endpoints enum type is not present, so the generated code is invalid:
```
@router.post(
    "/v1/whatever",
)
async def my_method_post(
    language_id: int = Query(None, description=""),
    dialog_start_type:  = Query(None, description="")
)
```
Note the `int `after `language_id` and lack of `DialogStartTypes `after `dialog_start_type`.
Is this also covered with your fix or should I open a separate bug?

**Comment by wing328** [comment]
please give `python-nextgen` client generator a try as it also leverages `pydantic`.

if the implementation, we can plot it to `python-fastapi`

**Comment by maxiptah** [comment]
`python-nextgen` is not listed, all I have is:
```
python-legacy
python
python-fastapi
python-prior
python-flask
python-aiohttp
python-blueplanet
```

**Comment by wing328** [comment]
are you using openapi-generator v6.3.0?

**Comment by maxiptah** [comment]
i'm using openapi-generator-cli as NPM package:
`"@openapitools/openapi-generator-cli": "^2.5.2"`

**Comment by wing328** [comment]
pls run `openapi-generator-cli version-manager set 6.3.0` to use v6.3.0

**Comment by maxiptah** [comment]
`python-nextgen` worked as I would expect, both problems are not there: enum type correctly passed to the controller, enum has members.
I would expect the same from `python-fastapi`, otherwise, it's unfortunately unusable..

File: modules/openapi-generator/src/main/resources/python-fastapi/model.mustache (hunk 1 above)
[the remark this answers is not in the document]
Comment (spacether) [reply modules/openapi-generator/src/main/resources/python-fastapi/model.mustache:29]: Enums can have any type. If type is unset enum values can be any json schema type

**Comment by Beaueve** [comment]
Any updates on this?

File: modules/openapi-generator/src/main/resources/python-fastapi/model.mustache (hunk 1 above)
[the remark this answers is not in the document]
Comment (Linus-Boehm) [reply modules/openapi-generator/src/main/resources/python-fastapi/model.mustache:29]: Can you provide an example from openapi?

**Comment by peyerroger** [comment]
Would be great if we had this fix in.

File: modules/openapi-generator/src/main/resources/python-fastapi/model.mustache (hunk 1 above)
[the remark this answers is not in the document]
Comment (spacether) [reply modules/openapi-generator/src/main/resources/python-fastapi/model.mustache:29]: Sure, here is an enum that contains a string, int, float, and boolean
```
EnumOfAnyType:
      enum:
        - "a"
        - 1
        - 3.14
        - true
```

[the remark this answers is not in the document]
Comment (peyerroger) [reply modules/openapi-generator/src/main/resources/python-fastapi/model.mustache:29]: Is this what we're looking for?

```
ReportingType:
  type: integer
  enum:
    - 1
    - 7
```

**Comment by peyerroger** [comment]
> please give `python-nextgen` client generator a try as it also leverages `pydantic`.
> 
> if the implementation looks good, we can plot it to `python-fastapi`

What would it mean to 'plot' it to `python-fastapi`. Could I somehow help?

**Comment by wing328** [comment]
Hi all,

I've filed https://github.com/OpenAPITools/openapi-generator/pull/16321 to improve pydantic support in python-fastapi generator (based on pydantic implementation in python client generator).

Can you please test the branch `python-fastapi-pydantic` and let me know how that goes?

My preliminary test result looks good but I would like more python-fastapi users to test it as I'm no expert in python-fastapi.

Thank you.

**Comment by wing328** [comment]
To test the branch, simply run the following:
```
git clone http://github.com/openapitools/openapi-generator -b python-fastapi-pydantic
cd python-fastapi-pydantic
mvn clean package -DskipTests -Dmaven.javadoc.skip=true
java -jar modules/openapi-generator-cli/target/openapi-generator-cli.jar generate -g python-fastapi -i modules/openapi-generator/src/test/resources/3_0/petstore.yaml -o /tmp/python-fastapi/
```

**Comment by jonasheschl** [comment]
Would it be possible to merge this onto master as a quickfix untill #16321 is ready? The current state of #16321 doesn't work our OpenAPI documents sadly. The generated code is invalid Python. With this PR merged onto master to get the changes from #14470 for inversion of control the generated FastAPI code works as expected for our OpenAPI documents.

**Comment by peyerroger** [comment]
Hi all.

I did some tests. The model classes look ok. The problem with the enums still persists in the router classes. List enums work, but single enums do not. It creates empty types:

```
openapi: 3.0.0
servers:
  - url: 'http://petstore.swagger.io/v2'
info:
  description: >-
    This is a sample server Petstore server. For this sample, you can use the api key
    `special-key` to test the authorization filters.
  version: 1.0.0
  title: OpenAPI Petstore
  license:
    name: Apache-2.0
    url: 'https://www.apache.org/licenses/LICENSE-2.0.html'
tags:
  - name: pet
    description: Everything about your Pets
  - name: store
    description: Access to Petstore orders
  - name: user
    description: Operations about user
paths:
  /fake/query_param_enum:
    get:
      tags:
        - fake
      summary: test query parameter enums
      description: ''
      operationId: fake_query_param_enum
      parameters:
        - name: stringEnum
          in: query
          description: Uses an enum query param.
          schema:
            type: string
            enum:
              - a
              - b
      responses:
        '400':
          description: Invalid username supplied
        '404':
          description: User not found

```

It creates something like that. (Note the type missing after 'status:'):

```
stringEnum:  = Query(None, description="Uses an enum query param"),
```

**Comment by jonasheschl** [comment]
> Hi all.
> 
> I did some tests. The model classes look ok. The problem with the enums still persists in the router classes. List enums work, but single enums do not. It creates empty types:
> 
> ```
> openapi: 3.0.0
> servers:
>   - url: 'http://petstore.swagger.io/v2'
> info:
>   description: >-
>     This is a sample server Petstore server. For this sample, you can use the api key
>     `special-key` to test the authorization filters.
>   version: 1.0.0
>   title: OpenAPI Petstore
>   license:
>     name: Apache-2.0
>     url: 'https://www.apache.org/licenses/LICENSE-2.0.html'
> tags:
>   - name: pet
>     description: Everything about your Pets
>   - name: store
>     description: Access to Petstore orders
>   - name: user
>     description: Operations about user
> paths:
>   /fake/query_param_enum:
>     get:
>       tags:
>         - fake
>       summary: test query parameter enums
>       description: ''
>       operationId: fake_query_param_enum
>       parameters:
>         - name: stringEnum
>           in: query
>           description: Uses an enum query param.
>           schema:
>             type: string
>             enum:
>               - a
>               - b
>       responses:
>         '400':
>           description: Invalid username supplied
>         '404':
>           description: User not found
> ```
> 
> It creates something like that. (Note the type missing after 'status:'):
> 
> ```
> stringEnum:  = Query(None, description="Uses an enum query param"),
> ```

You can try #16321. Some people have reported better results with that branch.

**Comment by jonasheschl** [comment]
Any update on this regarding a merge?

**Comment by wing328** [comment]
please help test https://github.com/OpenAPITools/openapi-generator/pull/17369 instead which uses pydantic v2

**Pull request closed without merging**